e2e/internal/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,378 @@
+//go:build e2e
+
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package upgrade
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/edgelesssys/constellation/v2/e2e/internal/kubectl"
+	"github.com/edgelesssys/constellation/v2/internal/config"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/semver"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type testFlags struct {
+	targetImage         string
+	targetKubernetes    string
+	targetMicroservices string
+	workspace           string
+	cliPath             string
+	wantWorker          int
+	wantControl         int
+}
+
+func parseFlags() testFlags {
+	targetImage = flag.String("target-image", "", "Image (shortversion) to upgrade to.")
+	targetKubernetes = flag.String("target-kubernetes", "", "Kubernetes version (MAJOR.MINOR.PATCH) to upgrade to. Defaults to default version of target CLI.")
+	targetMicroservices = flag.String("target-microservices", "", "Microservice version (MAJOR.MINOR.PATCH) to upgrade to. Defaults to default version of target CLI.")
+	// When executing the test as a bazel target the workspace path is supplied through an env variable that bazel sets.
+	workspace = flag.String("workspace", "", "Constellation workspace in which to run the tests.")
+	// When executing the test as a bazel target the CLI path is supplied through an env variable that bazel sets.
+	// When executing via `go test` extra care should be taken that the supplied CLI is built on the same commit as this test.
+	cliPath = flag.String("cli", "", "Constellation CLI to run the tests.")
+	wantWorker = flag.Int("want-worker", 0, "Number of wanted worker nodes.")
+	wantControl = flag.Int("want-control", 0, "Number of wanted control nodes.")
+
+	flag.Parse()
+
+	return testFlags{
+		targetImage:         *targetImage,
+		targetKubernetes:    *targetKubernetes,
+		targetMicroservices: *targetMicroservices,
+		workspace:           *workspace,
+		cliPath:             *cliPath,
+		wantWorker:          *wantWorker,
+		wantControl:         *wantControl,
+	}
+}
+
+// TestMain checks that the prerequisites for the test are met:
+// - a workspace is set
+// - a CLI path is set
+// - the constellation-upgrade folder does not exist
+func setup() (testFlags, error) {
```

Comment:
Yes this was previously a TestMain method. I switched to this style of defining and parsing the CLI flags based on katexochen's comment about removing global state. To address your concern about panics when calling this twice I added a check using flag's [Parsed](https://pkg.go.dev/flag#Parsed) and a comment that this should only be called once.

In general I would not want to add other tests to this file since it is already quite big and add separate pkgs/files for new e2e tests.

Hope that addresses your concern.